internal/tiltfile/k8s/locators.go

```diff
@@ -0,0 +1,117 @@
+package k8s
+
+import (
+	"fmt"
+
+	"go.starlark.net/starlark"
+
+	"github.com/tilt-dev/tilt/internal/k8s"
+	"github.com/tilt-dev/tilt/internal/tiltfile/value"
+)
+
+// Deserializing locators from starlark values.
+type JSONPathImageLocatorListSpec struct {
+	Specs []JSONPathImageLocatorSpec
+}
+
+func (s JSONPathImageLocatorListSpec) IsEmpty() bool {
+	return len(s.Specs) == 0
+}
+
+func (s *JSONPathImageLocatorListSpec) Unpack(v starlark.Value) error {
+	list := value.ValueOrSequenceToSlice(v)
+	for _, item := range list {
+		spec := JSONPathImageLocatorSpec{}
+		err := spec.Unpack(item)
+		if err != nil {
+			return err
+		}
+		s.Specs = append(s.Specs, spec)
+	}
+	return nil
+}
+
+func (s JSONPathImageLocatorListSpec) ToImageLocators(selector k8s.ObjectSelector) ([]k8s.ImageLocator, error) {
+	result := []k8s.ImageLocator{}
+	for _, spec := range s.Specs {
+		locator, err := spec.ToImageLocator(selector)
+		if err != nil {
+			return nil, err
+		}
+		result = append(result, locator)
+	}
+	return result, nil
+}
+
+type JSONPathImageLocatorSpec struct {
+	jsonPath string
+}
+
+func (s *JSONPathImageLocatorSpec) badValueErr(v starlark.Value) error {
+	return fmt.Errorf("Expected map of the form {'json_path': ..., 'repo_field': ..., 'tag_field': ...}. Actual: %s", v)
+}
+
+func (s *JSONPathImageLocatorSpec) Unpack(v starlark.Value) error {
+	var ok bool
+	s.jsonPath, ok = starlark.AsString(v)
+	if !ok {
+		return fmt.Errorf("Expected string, got: %s", v)
+	}
+	return nil
+}
+
+func (s JSONPathImageLocatorSpec) ToImageLocator(selector k8s.ObjectSelector) (k8s.ImageLocator, error) {
+	return k8s.NewJSONPathImageLocator(selector, s.jsonPath)
+}
+
+type JSONPathImageObjectLocatorSpec struct {
+	jsonPath  string
+	repoField string
+	tagField  string
+}
+
+func (s JSONPathImageObjectLocatorSpec) IsEmpty() bool {
+	return s == JSONPathImageObjectLocatorSpec{}
+}
+
+func (s *JSONPathImageObjectLocatorSpec) badValueErr(v starlark.Value) error {
+	return fmt.Errorf("Expected map of the form {'json_path': ..., 'repo_field': ..., 'tag_field': ...}. Actual: %s", v)
+}
+
+func (s *JSONPathImageObjectLocatorSpec) Unpack(v starlark.Value) error {
+	m, ok := v.(starlark.Mapping)
+	if !ok {
+		return s.badValueErr(v)
+	}
+
+	jsonPath, found, err := m.Get(starlark.String("json_path"))
+	if err != nil || !found {
+		return s.badValueErr(v)
+	}
+	repoField, found, err := m.Get(starlark.String("repo_field"))
+	if err != nil || !found {
+		return s.badValueErr(v)
+	}
+	tagField, found, err := m.Get(starlark.String("tag_field"))
+	if err != nil || !found {
+		return s.badValueErr(v)
+	}
+
+	s.jsonPath, ok = starlark.AsString(jsonPath)
+	if !ok {
+		return s.badValueErr(v)
```

Comment:
could you add a second arg to `badValueErr` with a more specific error? (e.g., "json_path must be a String. Found %T")

1. it'd annoying to play a guessing game as to which field is misspelled / which value is bad
2. the existing error message doesn't mention value types, so it wouldn't help with these errors anyway

Reply:
done